src/imitation/rewards/discrim_net.py

```diff
@@ -273,26 +273,76 @@ def _load(cls, directory):
     return cls(reward_net=reward_net, **params)
 
 
+DiscrimNetBuilder = Callable[[tf.Tensor, tf.Tensor],
+                             Tuple[util.LayersDict, tf.Tensor]]
+"""Type alias for function that takes an observation and action tensor and
+produces a tuple containing (1) a list of used TF layers, and (2) output logits
+for GAIL."""
+
+
+class build_mlp_discrim_net:
+  # this could be rewritten as a function that returns a closure instead of a
+  # functor class, but that would break pickle; instead I'm just pretending
+  # that build_mlp_discrim_net is a function that returns a closure :-)
+  def __init__(self, hid_sizes: Iterable[int] = (32, 32)):
+    """Construct a functor that builds a simple MLP-based discriminator for
+    GAIL. The returned function can be passed into the `build_discrim_net`
+    argument of `DiscrimNetGAIL`.
+
+    Args:
+        hid_sizes: list of layer sizes for each hidden layer of the network."""
+    self.hid_sizes = hid_sizes
+
+  def __call__(self, obs_input, act_input):
+    inputs = tf.concat([
+        tf.layers.flatten(obs_input),
+        tf.layers.flatten(act_input)], axis=1)
+
+    discrim_mlp = util.build_mlp(hid_sizes=self.hid_sizes)
+    discrim_logits = util.sequential(inputs, discrim_mlp)
+
+    return discrim_mlp, discrim_logits
+
+
 class DiscrimNetGAIL(DiscrimNet, serialize.LayersSerializable):
   """The discriminator to use for GAIL."""
 
-  def __init__(self,
-               observation_space: gym.Space,
-               action_space: gym.Space,
-               hid_sizes: Optional[Iterable[int]] = None,
-               scale: bool = False):
+  def __init__(
+        self,
+        observation_space: gym.Space,
+        action_space: gym.Space,
+        build_discrim_net: DiscrimNetBuilder = build_mlp_discrim_net(),
+        build_discrim_net_kwargs: Optional[dict] = None,
+        scale: bool = False):
+    """Construct discriminator network.
+
+    Args:
+        observation_space: observation space for this environment.
+        action_space: action space for this environment.
+        build_discrim_net: a callable that takes an observation input tensor
+            and action input tensor as input, then computes the logits
+            necessary to feed to GAIL. When called, the function should return
+            *both* a `LayersDict` containing all the layers used in
+            construction of the discriminator network, and a `tf.Tensor`
+            representing the desired discriminator logits.
+        build_discrim_net_kwargs: optional extra keyword arguments for
+            `build_discrim_net()`.
```

Comment:
Now that we have the `build_discrim_net_kwargs` pattern, could we make `build_mlp_discrim_net` into a function rather than a class? 

Not sure if `pytype` would be happy with Callable + arbitrary kwargs. (It looks like newer versions of Python will have a more precise `Protocol` type for defining callable types: https://stackoverflow.com/a/57840786/1091722)